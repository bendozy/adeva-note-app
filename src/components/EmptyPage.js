import React from 'react';
import PropTypes from 'prop-types';

const EmptyPage = ({ title, text }) => (
  <div>
    <h1>{title}</h1>
    <p>{text}</p>
  </div>
);

EmptyPage.defaultProps = {
  title: 'Page Not Found',
  text: 'The requested url could not be found',
};

EmptyPage.propTypes = {
  title: PropTypes.string,
  text: PropTypes.string,
};

export default EmptyPage;
