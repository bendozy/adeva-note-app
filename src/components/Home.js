import React from 'react';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  body: {
    marginTop: theme.spacing(5),
  },
}));

const Home = () => {
  const classes = useStyles();

  return (
    <Grid>
      <Grid>
        <Typography variant="h4" color="inherit" noWrap>
          Welcome
        </Typography>
      </Grid>
      <Grid>
        <Typography
          variant="body1"
          component="p"
          color="inherit"
          className={classes.body}
        >
          Hello there, thanks you for checking out this simple notes application.
          With this app, you can create, update, delete and search for notes.
        </Typography>
        <Typography
          variant="body1"
          component="p"
          color="inherit"
          className={classes.body}
        >
          This appliction uses localStorage for persistence. To get startted,
          click on the &quot;Add New Note&quot; button on the sidebar.
        </Typography>
      </Grid>
    </Grid>
  );
};

export default Home;
