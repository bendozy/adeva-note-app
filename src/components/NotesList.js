import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import InputBase from '@material-ui/core/InputBase';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

const useStyles = makeStyles({
  search: {
    marginTop: 5,
    display: 'flex',
    alignItems: 'center',
    width: '100%',
  },
  list: {
    // height: '93vh',
  },
  input: {
    marginLeft: 8,
    flex: 1,
  },
  iconButton: {
    padding: 10,
  },
});

const NotesList = ({ notes }) => {
  const classes = useStyles();
  const [searchText, setSearchText] = useState('');

  return (
    <Grid>
      <Grid>
        <Paper className={classes.search}>
          <IconButton className={classes.iconButton} aria-label="Menu">
            <MenuIcon />
          </IconButton>
          <InputBase
            className={classes.input}
            placeholder="Search Notes"
            inputProps={{ 'aria-label': 'Search Notes' }}
            value={searchText}
            onChange={event => setSearchText(event.target.value)}
          />
          <IconButton className={classes.iconButton} aria-label="Search">
            <SearchIcon />
          </IconButton>
        </Paper>
      </Grid>
      <Grid>
        <List component="nav" aria-label="Secondary mailbox folders">
          {notes
            .filter(note =>
              note.title.toLowerCase().includes(searchText.toLowerCase())
                || note.details.toLowerCase().includes(searchText.toLowerCase()))
            .map((note, index) => (
              <ListItem
                key={note.id}
                button
                divider={notes.length !== index + 1}
                component={Link}
                to={`/notes/${note.id}`}
              >
                <ListItemText primary={note.title} />
              </ListItem>
            ))
          }
        </List>
      </Grid>
      <Grid>
        <Button
          fullWidth
          variant="contained"
          color="primary"
          component={Link}
          to="/notes/add"
        >
          Add New Note
        </Button>
      </Grid>
    </Grid>
  );
};

NotesList.defaultProps = {
  notes: [],
};

NotesList.propTypes = {
  notes: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  })),
};

const mapStateToProps = ({ notes }) => ({ notes });

export default connect(mapStateToProps)(NotesList);
