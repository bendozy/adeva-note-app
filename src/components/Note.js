import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import classNames from 'classnames';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import Fab from '@material-ui/core/Fab';
import Icon from '@material-ui/core/Icon';
import DeleteIcon from '@material-ui/icons/Delete';
import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import EmptyPage from './EmptyPage';
import { deleteNote as deleteNoteAction } from '../actions/notes';

const useStyles = makeStyles(theme => ({
  details: {
    marginTop: theme.spacing(5),
  },
  fab: {
    margin: theme.spacing(1),
    position: 'fixed',
    bottom: theme.spacing(2),
    right: theme.spacing(2),
  },
  fabEdit: {
    right: theme.spacing(10),
  },
  fabDelete: {
    right: theme.spacing(2),
  },
}));

const Note = ({
  history,
  deleteNoteAction: deleteNote,
  note,
  note: {
    id,
    title,
    details,
  } = {},
}) => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const [snackOpen, setSnackOpen] = useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const handleSubmit = () => {
    deleteNote(note.id);
    setOpen(false);
    setSnackOpen(true);
    history.push('/');
  };

  if (!note) {
    return (
      <EmptyPage
        title="Note Not Found"
        text="The requested note is not available"
      />
    );
  }

  return (
    <React.Fragment>
      <Grid>
        <Grid>
          <Typography variant="h4" color="inherit" noWrap>
            {title}
          </Typography>
        </Grid>
        <Grid>
          <Typography
            variant="body1"
            component="p"
            color="inherit"
            className={classes.details}
          >
            {details}
          </Typography>
        </Grid>
      </Grid>
      <Fab
        color="primary"
        aria-label="Edit Note"
        className={classNames(classes.fab, classes.fabEdit)}
        component={Link}
        to={`/notes/${id}/edit`}
      >
        <Icon>edit_icon</Icon>
      </Fab>
      <Fab
        color="secondary"
        aria-label="Delete Note"
        className={classNames(classes.fab, classes.fabDelete)}
        onClick={handleOpen}
      >
        <DeleteIcon />
      </Fab>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">Delete Note?</DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure you want to delete this note.
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button onClick={handleSubmit} color="secondary" autoFocus>
            Yes, Delete
          </Button>
        </DialogActions>
      </Dialog>
      <Snackbar
        anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
        open={snackOpen}
        onClose={() => setSnackOpen(false)}
        autoHideDuration={4000}
        ContentProps={{
          'aria-describedby': 'message-id',
        }}
        message={<span id="message-id">Note deleted successfully</span>}
      />
    </React.Fragment>
  );
};

Note.defaultProps = {
  note: undefined,
};

Note.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  }),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  deleteNoteAction: PropTypes.func.isRequired,
};

const mapStateToProps = ({ notes }, { match: { params } }) =>
  ({
    note: notes.find(note => note.id === parseInt(params.noteId, 10)),
  });

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      deleteNoteAction,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(Note);
