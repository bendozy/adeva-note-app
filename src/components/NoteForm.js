import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import EmptyPage from './EmptyPage';
import {
  createNote as createNoteAction,
  updateNote as updateNoteAction,
} from '../actions/notes';

const useStyles = makeStyles(theme => ({
  form: {
    width: '100%',
    marginTop: theme.spacing(1),
  },
}));

export const NoteForm = ({
  note,
  nextId,
  history,
  isEditMode,
  createNoteAction: createNote,
  updateNoteAction: updateNote,
}) => {
  const classes = useStyles();
  const [values, setValues] = useState({
    title: note ? note.title : '',
    details: note ? note.details : '',
  });
  const [errors, setErrors] = useState({});

  const handleChange = name => (event) => {
    setValues({ ...values, [name]: event.target.value });
    setErrors({ ...errors, [name]: undefined });
  };

  const formatInput = name => (event) => {
    setValues({ ...values, [name]: event.target.value.trim() });
  };

  const validate = () => {
    let isValid = true;
    const formErrors = {};

    setErrors({});

    if (!values.title) {
      formErrors.title = 'Enter Note Title';
      isValid = false;
    }

    if (!values.details) {
      formErrors.details = 'Enter Note Details';
      isValid = false;
    }

    setErrors(formErrors);

    return isValid;
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (validate()) {
      if (note) {
        updateNote(Object.assign({}, note, values));
        history.push(`/notes/${note.id}`);
      } else {
        createNote(values);
        history.push(`/notes/${nextId}`);
      }
    }
  };

  if (isEditMode && !note) {
    return (
      <EmptyPage
        title="Note Not Found"
        text="The requested note is not available"
      />
    );
  }

  return (
    <Grid>
      <Grid>
        <Typography variant="h4" color="inherit" noWrap>
          {note ? 'Update' : 'New'} Note
        </Typography>
      </Grid>
      <form
        className={classes.form}
        onSubmit={handleSubmit}
        noValidate
      >
        <Grid>
          <TextField
            required
            id="title"
            label="Note Title"
            value={values.title}
            fullWidth
            onChange={handleChange('title')}
            onBlur={formatInput('title')}
            margin="normal"
            error={errors.title}
          />
        </Grid>
        <Grid>
          <TextField
            required
            id="title"
            label="Note Details"
            value={values.details}
            fullWidth
            onChange={handleChange('details')}
            multiline
            rows={20}
            margin="normal"
            error={errors.details}
          />
        </Grid>
        <Grid>
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            {note ? 'Update' : 'Create'} Note
          </Button>
        </Grid>
      </form>
    </Grid>
  );
};

NoteForm.defaultProps = {
  note: undefined,
};

NoteForm.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  }),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  nextId: PropTypes.number.isRequired,
  isEditMode: PropTypes.bool.isRequired,
  createNoteAction: PropTypes.func.isRequired,
  updateNoteAction: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ notes, nextId }, { match: { params } }) =>
  ({
    note: notes.find(note => note.id === parseInt(params.noteId, 10)),
    isEditMode: !!params.noteId,
    nextId,
  });

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      createNoteAction,
      updateNoteAction,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(NoteForm);
