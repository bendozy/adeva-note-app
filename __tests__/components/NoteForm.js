import React from 'react';
import toJson from 'enzyme-to-json';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import EmptyPage from '../../src/components//EmptyPage';
import { NoteForm, mapStateToProps, mapDispatchToProps } from '../../src/components/NoteForm';

describe('<NoteForm />', () => {
  it('should render a <Note /> component', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode={false}
      nextId={2}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(Typography)).toHaveLength(1);
    expect(wrapper.find(TextField)).toHaveLength(2);
    expect(wrapper.find(Button)).toHaveLength(1);
  });

  it('should render a <EmptyPage /> component when in Edit mode with no note', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode
      nextId={2}
    />);

    expect(wrapper.find(EmptyPage)).toHaveLength(1);
    expect(wrapper.find(EmptyPage).first().props().title).toBe('Note Not Found');
    expect(wrapper.find(EmptyPage).first().props().text).toBe('The requested note is not available');
  });

  it('should render a form with initial values when in Edit mode with note', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode
      nextId={2}
      note={{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }}
    />);

    expect(wrapper.find(TextField)).toHaveLength(2);
    expect(wrapper.find(Typography)).toHaveLength(1);
    expect(wrapper.find(TextField).first().props().value).toBe('Test Example');
    expect(wrapper.find(TextField).at(1).props().value).toBe('Testing detail');
  });

  it('should handle change events', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode={false}
      nextId={2}
    />);

    wrapper.find(TextField).first().simulate('change', { target: { value: 'dope title' } });
    wrapper.find(TextField).at(1).simulate('change', { target: { value: 'dope detail' } });

    expect(wrapper.find(TextField).first().props().value).toBe('dope title');
    expect(wrapper.find(TextField).at(1).props().value).toBe('dope detail');
  });

  it('should handle blur events', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode={false}
      nextId={2}
    />);

    wrapper.find(TextField).first().simulate('change', { target: { value: '   title' } });
    wrapper.find(TextField).first().simulate('blur', { target: { value: '   title' } });

    expect(wrapper.find(TextField).first().props().value).toBe('title');
  });

  it('should show validation errors if required fields are not filled', () => {
    const wrapper = shallow(<NoteForm
      createNoteAction={jest.fn()}
      updateNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      isEditMode={false}
      nextId={2}
    />);

    wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(wrapper.find(TextField).first().props().error).toBe('Enter Note Title');
    expect(wrapper.find(TextField).at(1).props().error).toBe('Enter Note Details');
  });

  it('should create note if required fields have values', () => {
    const createNoteAction = jest.fn();
    const updateNoteAction = jest.fn();

    const wrapper = shallow(<NoteForm
      createNoteAction={createNoteAction}
      updateNoteAction={updateNoteAction}
      history={{
        push: jest.fn(),
      }}
      isEditMode={false}
      nextId={2}
    />);

    wrapper.find(TextField).first().simulate('change', { target: { value: 'My new title' } });
    wrapper.find(TextField).at(1).simulate('change', { target: { value: 'My new details' } });
    wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(createNoteAction).toHaveBeenCalled();
    expect(updateNoteAction).not.toHaveBeenCalled();
  });

  it('should update note if required fields have values', () => {
    const createNoteAction = jest.fn();
    const updateNoteAction = jest.fn();

    const wrapper = shallow(<NoteForm
      createNoteAction={createNoteAction}
      updateNoteAction={updateNoteAction}
      history={{
        push: jest.fn(),
      }}
      note={{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }}
      isEditMode
      nextId={2}
    />);

    wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(createNoteAction).not.toHaveBeenCalled();
    expect(updateNoteAction).toHaveBeenCalled();
  });

  it('check for the correct note title and editMode', () => {
    const mapStatFunc = mapStateToProps({
      notes: [{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }],
    }, {
      match: {
        params: {
          noteId: 1,
        },
      },
    });

    expect(mapStatFunc.note.title).toBe('Test Example');
    expect(mapStatFunc.isEditMode).toBe(true);
  });

  it('should discpatch createNote action', () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).createNoteAction();
    expect(dispatch.mock.calls[0][0].type).toEqual('CREATE_NOTE');
  });

  it('should discpatch updateNote action', () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).updateNoteAction();
    expect(dispatch.mock.calls[0][0].type).toEqual('UPDATE_NOTE');
  });
});
