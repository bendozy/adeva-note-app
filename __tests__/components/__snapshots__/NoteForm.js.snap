// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteForm /> should render a <Note /> component 1`] = `
<WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Typography))
      color="inherit"
      noWrap={true}
      variant="h4"
    >
      New
       Note
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Grid))>
  <form
    className="makeStyles-form-1"
    noValidate={true}
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="title"
        label="Note Title"
        margin="normal"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        value=""
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="title"
        label="Note Details"
        margin="normal"
        multiline={true}
        onChange={[Function]}
        required={true}
        rows={20}
        value=""
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        fullWidth={true}
        type="submit"
        variant="contained"
      >
        Create
         Note
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </form>
</WithStyles(ForwardRef(Grid))>
`;
